From a2bc58abbae8e8e08d9f0e10d77ef1bb9df3f8ac Mon Sep 17 00:00:00 2001
From: Stefan Lange <s.lange@gateware.de>
Date: Tue, 29 Nov 2016 11:50:59 +0100
Subject: [PATCH] DP83867 ethernet phy driver: add clock output select

Enhance DP83867 ethernet phy driver to be able to select the clock
output for the CLK_OUT pin via device tree property.

Signed-off-by: Stefan Lange <s.lange@gateware.de>
---
 drivers/net/phy/dp83867.c            | 19 +++++++++++++++++++
 include/dt-bindings/net/ti-dp83867.h | 14 ++++++++++++++
 2 files changed, 33 insertions(+)

diff --git a/drivers/net/phy/dp83867.c b/drivers/net/phy/dp83867.c
index 3fe8e45..67ffc49 100644
--- a/drivers/net/phy/dp83867.c
+++ b/drivers/net/phy/dp83867.c
@@ -33,6 +33,7 @@
 /* Extended Registers */
 #define DP83867_RGMIICTL	0x0032
 #define DP83867_RGMIIDCTL	0x0086
+#define DP83867_IOMUXCFG	0x0170
 
 #define DP83867_SW_RESET	BIT(15)
 #define DP83867_SW_RESTART	BIT(14)
@@ -61,10 +62,15 @@
 /* RGMIIDCTL bits */
 #define DP83867_RGMII_TX_CLK_DELAY_SHIFT	4
 
+/* IOMUXCFG bits */
+#define DP83867_IOMUXCFG_CLK_O_SEL_SHIFT        8
+#define DP83867_IOMUXCFG_CLK_O_SEL_MASK         0x1F00
+
 struct dp83867_private {
 	int rx_id_delay;
 	int tx_id_delay;
 	int fifo_depth;
+	int clk_out_sel;
 	int values_are_sane;
 };
 
@@ -129,6 +135,11 @@ static int dp83867_of_init(struct phy_device *phydev)
 	if (ret)
 		goto invalid_dt;
 
+        ret = of_property_read_u32(of_node, "ti,dp83867-clk-out-sel",
+                                   &dp83867->clk_out_sel);
+        if (ret)
+                goto invalid_dt;
+
 	dp83867->values_are_sane = 1;
 
 	return 0;
@@ -209,6 +220,14 @@ static int dp83867_config_init(struct phy_device *phydev)
 				       DP83867_DEVADDR, phydev->addr, delay);
 	}
 
+	/* set Clock Output Select in IO_MUX_CFG register */
+	val = phy_read_mmd_indirect(phydev, DP83867_IOMUXCFG,
+                                    DP83867_DEVADDR, phydev->addr);
+	val &= ~DP83867_IOMUXCFG_CLK_O_SEL_MASK;
+	val |= (dp83867->clk_out_sel << DP83867_IOMUXCFG_CLK_O_SEL_SHIFT);
+        phy_write_mmd_indirect(phydev, DP83867_IOMUXCFG,
+                               DP83867_DEVADDR, phydev->addr, val);
+
 	return 0;
 }
 
diff --git a/include/dt-bindings/net/ti-dp83867.h b/include/dt-bindings/net/ti-dp83867.h
index 172744a..c0d057d 100644
--- a/include/dt-bindings/net/ti-dp83867.h
+++ b/include/dt-bindings/net/ti-dp83867.h
@@ -42,4 +42,18 @@
 #define	DP83867_RGMIIDCTL_3_75_NS	0xe
 #define	DP83867_RGMIIDCTL_4_00_NS	0xf
 
+#define DP83867_IOMUXCFG_CLK_O_SEL_REF_CLK      0x0C
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHD_TX       0x0B
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHC_TX       0x0A
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHB_TX       0x09
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHA_TX       0x08
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHD_RX_DIV5  0x07
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHC_RX_DIV5  0x06
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHB_RX_DIV5  0x05
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHA_RX_DIV5  0x04
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHD_RX       0x03
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHC_RX       0x02
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHB_RX       0x01
+#define DP83867_IOMUXCFG_CLK_O_SEL_CHA_RX       0x00
+
 #endif
-- 
1.9.1

